TYPO3.TypoScript/Classes/TYPO3/TypoScript/Core/Cache/CacheSegmentParser.php

```diff
@@ -21,134 +22,241 @@ class CacheSegmentParser
+     * - cleanContent -> the raw content without any cache references for this segment and all sub segments
+     * - embed -> the placeholder content for this segment to be used in "content" of parent segments
+     *
+     * @param integer $currentPosition
+     * @return array
+     * @throws Exception
+     */
+    protected function parseSegment($currentPosition)
+    {
+        $nextStartPosition = $this->calculateNextTokenPosition($currentPosition, ContentCache::CACHE_SEGMENT_START_TOKEN);
+        if ($nextStartPosition !== $currentPosition) {
+            throw new Exception(sprintf('The current position (%d) is not the start of a segment, next start position %d', $currentPosition, $nextStartPosition), 1472464124);
+        }
+
+        $segmentData = [
+            'identifier' => '',
+            'type' => '',
+            'context' => '',
+            'metadata' => '',
+            'content' => '',
+            'cleanContent' => '',
+            'embed' => ''
+,        ];
+
+        $nextEndPosition = $this->calculateNextTokenPosition($currentPosition, ContentCache::CACHE_SEGMENT_END_TOKEN);
+        $currentPosition = $this->calculateCurrentPosition($nextStartPosition);
+        $nextStartPosition = $this->calculateNextTokenPosition($currentPosition, ContentCache::CACHE_SEGMENT_START_TOKEN);
+
+        $nextIdentifierSeparatorPosition = $this->calculateNextTokenPosition($currentPosition, ContentCache::CACHE_SEGMENT_SEPARATOR_TOKEN);
+        $nextSecondIdentifierSeparatorPosition = $this->calculateNextTokenPosition($nextIdentifierSeparatorPosition + 1, ContentCache::CACHE_SEGMENT_SEPARATOR_TOKEN);
+
+        if ($nextIdentifierSeparatorPosition === false || $nextSecondIdentifierSeparatorPosition === false
+            || $nextStartPosition !== false && $nextStartPosition < $nextIdentifierSeparatorPosition
+            || $nextEndPosition !== false && $nextEndPosition < $nextIdentifierSeparatorPosition
+            || $nextStartPosition !== false && $nextStartPosition < $nextSecondIdentifierSeparatorPosition
+            || $nextEndPosition !== false && $nextEndPosition < $nextSecondIdentifierSeparatorPosition
+        ) {
+            throw new Exception(sprintf('Missing segment separator token after position %d', $currentPosition), 1391855139);
+        }
+
+        $identifier = $this->extractContent($currentPosition, $nextIdentifierSeparatorPosition);
+        $contextOrMetadata = $this->extractContent($this->calculateCurrentPosition($nextIdentifierSeparatorPosition), $nextSecondIdentifierSeparatorPosition);
+
+        $segmentData['identifier'] = $identifier;
+        if (strpos($identifier, 'eval=') === 0) {
+            $segmentData['type'] = ContentCache::SEGMENT_TYPE_UNCACHED;
+            $segmentData['context'] = $contextOrMetadata;
+            $this->uncachedPartCount++;
+        } elseif (strpos($identifier, 'evalCached=') === 0) {
+            $segmentData['type'] = ContentCache::SEGMENT_TYPE_DYNAMICCACHED;
```

Comment:
Done in https://github.com/neos/neos-development-collection/pull/683/commits/a6c3dacb5a79e2723c158fc06fa06403d7b3cb17